### PR TITLE
Update enum_av.py

### PR DESCRIPTION
**Pull request by bongobongoland**
Works slightly faster.

### CODE REVIEW

**Comment by Marshall-Hallenbeck** [comment]
@bongobongoland which OS versions and with which AV/EDRs did you test this with?

**Comment by bongobongoland** [comment]
2008, 2012, 2016, 2019
ESET
Kaspersky
Defender
Acronis
Sophos

**Comment by bongobongoland** [comment]
I have fixed the bug where the module wouldn't inform about non-detection.

File: cme/modules/enum_av.py [hunk 1]
```diff
@@ -3,106 +3,102 @@
...
     multiple_hosts = True
 
     def __init__(self, context=None, module_options=None):
         self.context = context
         self.module_options = module_options
-
+        
     def options(self, context, module_options):
-        """ """
+        """Placeholder for module options."""
```
Comment (Marshall-Hallenbeck) [inline cme/modules/enum_av.py:25]: Fill this in?

File: cme/modules/enum_av.py [hunk 2]
```diff
@@ -3,106 +3,102 @@
...
-
 class CMEModule:
-    """
-    Uses LsarLookupNames and NamedPipes to gather information on all endpoint protection solutions installed on the the remote host(s)
-    Module by @mpgn_x64
-    """
-
+    
     name = "enum_av"
-    description = "Gathers information on all endpoint protection solutions installed on the the remote host(s) via LsarLookupNames (no privilege needed)"
```
Comment (Marshall-Hallenbeck) [inline cme/modules/enum_av.py:20]: AV/EDR is endpoint protection, why change the wording? also why remove the info about the method used?

File: cme/modules/enum_av.py [hunk 3]
```diff
@@ -3,106 +3,102 @@
...
+                    except:
                         pass
-            success += 1
+
         except Exception as e:
             context.log.fail(str(e))
 
+        return results
+
+    def _detect_running_processes(self, context, connection, results):
```
Comment (Marshall-Hallenbeck) [inline cme/modules/enum_av.py:73]: Why is this tagged as a private method (underscore)?

**Review by Marshall-Hallenbeck — COMMENTED** [review]
Can you give some info on how this is slightly faster and by how much?

File: cme/modules/enum_av.py (hunk 1 above)
Comment (bongobongoland) [reply cme/modules/enum_av.py:25]: there are no options at this point but I might introduces a few in the future.

File: cme/modules/enum_av.py (hunk 2 above)
Comment (bongobongoland) [reply cme/modules/enum_av.py:20]: Got butchered when I was messing with the code. I typed it in manually.

File: cme/modules/enum_av.py (hunk 3 above)
Comment (bongobongoland) [reply cme/modules/enum_av.py:73]: encapsulation for future use :)

**Comment by bongobongoland** [comment]
> Can you give some info on how this is slightly faster and by how much?
Marginally faster. 10% or so. We still have to rely on smb which is the bottleneck. @XiaoliChan will try to improve the speed ;-)

**Comment by XiaoliChan** [comment]
> Marginally faster. 10% or so. We still have to rely on smb which is the bottleneck. @XiaoliChan will try to improve the speed ;-)

HMM, I also have no idea how to improve the speed, because I haven't use it before LOL

**Comment by bongobongoland** [comment]
> > Marginally faster. 10% or so. We still have to rely on smb which is the bottleneck. @XiaoliChan will try to improve the speed ;-)
> 
> HMM, I also have no idea how to improve the speed, because I haven't use it before LOL

@XiaoliChan I mean you have ideas how to improve the speed of SMB

**Comment by XiaoliChan** [comment]
> @XiaoliChan I mean you have ideas how to improve the speed of SMB

XD

**Comment by bongobongoland** [comment]
@XiaoliChan I believe in you!

**Comment by XiaoliChan** [comment]
> @XiaoliChan I believe in you!

Don't think we can do improvement in SMB connection, it's really heavy.

**Comment by XiaoliChan** [comment]
Hey, can you also rename the variable “k” to "doKerberos", this is better for human readable.

File: cme/modules/enum_av.py [hunk 4]
```diff
@@ -3,106 +3,102 @@
...
-                connection.kerberos,
-                connection.domain,
-                connection.lmhash,
-                connection.nthash,
-                connection.aesKey,
+                domain=connection.domain,
+                username=connection.username,
+                password=connection.password,
+                remote_name=target,
+                k=connection.kerberos,
```
Comment (XiaoliChan) [inline cme/modules/enum_av.py:49]: maybe "doKerberos" is better than "k"

**Comment by bongobongoland** [comment]
I'm working on the module to speed things up a bit more. It's still too slow.

**Comment by bongobongoland** [comment]
Done.

**Comment by bongobongoland** [comment]
> > @XiaoliChan I believe in you!
> 
> Don't think we can do improvement in SMB connection, it's really heavy.

I still believe in you. No matter what.

File: cme/modules/enum_av.py [hunk 5]
```diff
@@ -2,20 +2,16 @@
...
 from impacket.dcerpc.v5.dtypes import NULL, MAXIMUM_ALLOWED, RPC_UNICODE_STRING
-from impacket.dcerpc.v5 import transport
 import pathlib
 
-
 class CMEModule:
-    """
-    Uses LsarLookupNames and NamedPipes to gather information on all endpoint protection solutions installed on the the remote host(s)
-    Module by @mpgn_x64
-    """
```
Comment (NeffIsBack) [inline cme/modules/enum_av.py:17]: Why remove the Class description?

Comment (bongobongoland) [reply cme/modules/enum_av.py]: what's its purpose? :)

Comment (NeffIsBack) [reply cme/modules/enum_av.py]: It should tell people looking at the Code what the Module does and who made it :)

**Comment by bongobongoland** [comment]
Shouldn't it be called not class CMEModule but NXCModule? :)

**Comment by XiaoliChan** [comment]
> Shouldn't it be called not class CMEModule but NXCModule? :)

Yeah `NXCModule` is right

**Comment by Dfte** [comment]
About the AV/EDR detection, I know that drivers rely on an altitude nulmber which is an integer value that identifies a driver and whom it belongs to: https://learn.microsoft.com/en-us/windows-hardware/drivers/ifs/allocated-altitudes#320000---329998-fsfilter-anti-virus

Looking at the range 320000 - 3299998 we would be able to list installed drivers only by requesting a registry key.
What do you think about that ?

**Comment by bongobongoland** [comment]
> About the AV/EDR detection, I know that drivers rely on an altitude nulmber which is an integer value that identifies a driver and whom it belongs to: https://learn.microsoft.com/en-us/windows-hardware/drivers/ifs/allocated-altitudes#320000---329998-fsfilter-anti-virus
> 
> Looking at the range 320000 - 3299998 we would be able to list installed drivers only by requesting a registry key. What do you think about that ?

I think it's a brilliant idea. I need to look into it. Because checking for a driver might be significantly slower than checking for a certain service.

**Comment by bongobongoland** [comment]
Something I quickly noticed, some products are not in the list. For instance, CyNetDrv.sys from cynet isn't there.

**Comment by Dfte** [comment]
> Something I quickly noticed, some products are not in the list. For instance, CyNetDrv.sys from cynet isn't there.

Oh wow so that means the list is not updated! I'll take a look at it also :)

**Comment by bongobongoland** [comment]
cylance is a different company :) 
Cynet has 2 drivers: cdrv.sys and CyNetDrv.sys
Only minifilter drivers have altitude numbers.

**Comment by Dfte** [comment]
Oh my bad, I wasn't fully awoken xD
Ok so altitude numbers are not useful unfortunately.

**Comment by bongobongoland** [comment]
It is useful. Thanks for bringing it up. So far I can only see one product missing from the list. Other AVs\EDRs are pretty much all there.

**Comment by bongobongoland** [comment]
For the time being, we'll keep the current LsarNameLookup method in place. But I'll keep the altitude method in mind.

**Comment by NeffIsBack** [comment]
> Yeah but if you can list running services, there is no need of checking the altitude number aswell. That would just create more IOC's as we'll list both remote services and registry keys. Is it useful to list altitude numbers if you already can list services more accurately ?

The main problem i see with getting the altitude numbers is it has to be queried from the registry right? If so this would require admin privs and therefore kinda defeat the purpose of the module. If there is a way to query these values it would be a really good addition imo

**Comment by bongobongoland** [comment]
Yes. that is true. You need admin rights. Personally, I have never ever used the module with lower privileges. Always use it as admin. 
I like the altitude method because of two things:
1. it's more universal. in most cases you need only one number to identify a product regardless of its version. with services this can be tricky like in Cisco's EDR where every service has a version number in it's name (I did find a way to identify the product using services tho). You can't use LNL method to lookup partially. 

3. it's faster than LNL. :)

**Comment by NeffIsBack** [comment]
@bongobongoland can you merge in the dev branch and add your changes to the new file in /nxc/modules/enum_av?

**Comment by bongobongoland** [comment]
done. I think :)

**Comment by NeffIsBack** [comment]
> done. I think :)

Did you push the changes? Looks to me as there werent any new commits

**Comment by bongobongoland** [comment]
"[bongobongoland](https://github.com/bongobongoland) wants to merge 5 commits into [Pennyw0rth:develop](https://github.com/Pennyw0rth/NetExec/tree/develop) from [bongobongoland:main](https://github.com/bongobongoland/NetExec/tree/main)"

is that correct?

**Pull request closed without merging**